fiftyone/core/fields.py

```diff
@@ -600,12 +614,110 @@ class EmbeddedDocumentField(mongoengine.fields.EmbeddedDocumentField, Field):
 
+    def validate(self, value, clean=True, expand=False):
+        if self._validation_schema is None:
+            self._validation_schema = self.get_field_schema()
+
+        schema = self._validation_schema
+        for name in value._fields_ordered:
+            field = None
+            field_value = value.get_field(name)
+            if field_value is None:
+                continue
+
+            if name.startswith("_"):
+                continue
+
+            field = schema.get(name, None)
+            if field is None and expand:
+                field_kwargs = foo.get_implied_field_kwargs(field_value)
+                field = foo.create_field(name, **field_kwargs)
+                self._save_field(field, [name])
+            elif field is None:
+                self.error("field does not have field '%s' declared" % name)
+            elif isinstance(field, EmbeddedDocumentField):
+                field.validate(field_value, clean=False, expand=expand)
+            else:
+                field.validate(field_value)
+
+        if isinstance(value, foo.DynamicEmbeddedDocument):
+            value._set_parent(self)
+
+        super().validate(value, clean)
+
+    def get_field_schema(
+        self, ftype=None, embedded_doc_type=None, include_private=False
+    ):
+        """Returns a schema dictionary describing the fields of the embedded
+        document field.
+
+        Args:
+            ftype (None): an optional field type to which to restrict the
+                returned schema. Must be a subclass of
+                :class:`fiftyone.core.fields.Field`
+            embedded_doc_type (None): an optional embedded document type to
+                which to restrict the returned schema. Must be a subclass of
+                :class:`fiftyone.core.odm.BaseEmbeddedDocument`
+            include_private (False): whether to include fields that start with
+                ``_`` in the returned schema
+
+        Returns:
+             an ``OrderedDict`` mapping field names to field types
+        """
+        fields = {
+            name: field for name, field in self.document_type._fields.items()
+        }
+        fields.update({field.name: field for field in self.fields})
+
+        filtered_fields = {}
+
+        for name, field in fields.items():
+            if not include_private and name.startswith("_"):
+                continue
+
+            if ftype and not isinstance(field, ftype):
+                continue
+
+            if embedded_doc_type and (
+                not isinstance(field, EmbeddedDocumentField)
+                or embedded_doc_type != field.document_type
+            ):
+                continue
+
+            filtered_fields[name] = field
+
+        return filtered_fields
+
+    def to_python(self, value):
+        doc = super().to_python(value)
+
+        if isinstance(doc, foo.DynamicEmbeddedDocument):
+            doc._set_parent(self)
+        return doc
+
+    def _save_field(self, field, keys):
```

Comment:
The `_parent` of an `EmbeddedDocumentField` is either a sample or frame backing document **class** or an `EmbeddedDocumentField` **instance**.

frame and sample document backing classes have an equivalent `_save_field` method, but it is a `@classmethod`.

These `_save_field()` methods propagate schema changes back up to the dataset document for saving.